### PR TITLE
[TS Migration] Migrate storage, utils and OnyxCache to Typescript

### PR DESCRIPTION
**Pull request by blazejkustra**
### Related Issues

GH_LINK

### Author Checklist

- [ ] I linked the correct issue in the `### Related Issues` section above
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: lib/utils.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            const shouldOmitSourceKey = shouldRemoveNullObjectValues && isSourceOrTargetNull;
+
+            if (!shouldOmitSourceKey) {
+                destination[key] = targetValue;
+            }
+        }
+    }
+
+    // lodash adds a small overhead so we don't use it here
+    // eslint-disable-next-line rulesdir/prefer-underscore-method
```
Comment (BartoszGrajdek) [inline lib/utils.ts:39]: Maybe we could remove this rule? At least for TS files

**Review by BartoszGrajdek — APPROVED** [review]

File: lib/OnyxCache.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * Encapsulates Onyx cache related functionality
+ */
+class OnyxCache {
+    private storageKeys: Set<string>;
+
+    private recentKeys: Set<string>;
+
+    private storageMap: StorageMap;
+
+    private pendingPromises: Map<string, Promise<unknown>>;
```
Comment (fabioh8010) [inline lib/OnyxCache.ts:17]: Could you add back the comments for these properties?

File: lib/OnyxCache.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    private maxRecentKeysSize = 0;
+
+    constructor() {
+        this.storageKeys = new Set();
+        this.recentKeys = new Set();
+        this.storageMap = {};
+        this.pendingPromises = new Map();
```
Comment (fabioh8010) [inline lib/OnyxCache.ts:25]: Shouldn't we keep the bindAll thing?

```ts
        // bind all public methods to prevent problems with `this`
        _.bindAll(
            this,
            'getAllKeys',
            'getValue',
            'hasCacheForKey',
            'addKey',
            'set',
            'drop',
            'merge',
            'hasPendingTask',
            'getTaskPromise',
            'captureTask',
            'removeLeastRecentlyUsedKeys',
            'setRecentKeysLimit',
        );
```

File: lib/OnyxCache.ts
Comment (fabioh8010) [inline lib/OnyxCache.ts:1]: Could you add back the comments of all functions here?

Comment (fabioh8010) [reply lib/OnyxCache.ts:1]: Is there a reason for using `string` and `unknown` instead of `OnyxKey` and `OnyxValue` types in this file?

File: lib/storage/WebStorage.ts [hunk 4]
```diff
@@ -3,32 +3,28 @@
...
 
 /**
  * Raise an event thorough `localStorage` to let other tabs know a value changed
- * @param {String} onyxKey
  */
-function raiseStorageSyncEvent(onyxKey) {
+function raiseStorageSyncEvent(onyxKey: Key) {
     global.localStorage.setItem(SYNC_ONYX, onyxKey);
-    global.localStorage.removeItem(SYNC_ONYX, onyxKey);
+    global.localStorage.removeItem(SYNC_ONYX);
```
Comment (fabioh8010) [inline lib/storage/WebStorage.ts:17]: Why this change?

File: lib/storage/providers/IDBKeyVal.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    getDatabaseSize() {
+        if (!window.navigator || !window.navigator.storage) {
+            throw new Error('StorageManager browser API unavailable');
+        }
+
+        return window.navigator.storage
+            .estimate()
+            .then((value) => ({
+                bytesUsed: value.usage || 0,
+                bytesRemaining: (value.quota || 0) - (value.usage || 0),
```
Comment (fabioh8010) [inline lib/storage/providers/IDBKeyVal.ts:60]:
```suggestion
                bytesUsed: value.usage ?? 0,
                bytesRemaining: (value.quota ?? 0) - (value.usage ?? 0),
```

wdyt?

File: lib/storage/providers/SQLiteStorage.ts [hunk 6]
```diff
@@ -85,84 +58,43 @@ const provider = {
...
                 bytesUsed: pageSize * pageCount,
                 bytesRemaining,
             };
         });
     },
-
-    /**
-     * Noop on native
-     */
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
```
Comment (fabioh8010) [inline lib/storage/providers/SQLiteStorage.ts:97]:
```suggestion

    // eslint-disable-next-line @typescript-eslint/no-empty-function
```

nit

File: lib/storage/providers/types.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import type {BatchQueryResult, QueryResult} from 'react-native-quick-sqlite';
+
+type Key = string;
+type Value = IDBValidKey;
```
Comment (fabioh8010) [inline lib/storage/providers/types.ts:4]: Where does `IDBValidKey` type come from?

File: lib/storage/providers/types.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    setMemoryOnlyKeys: () => void;
+
+    /**
+     * Gets the total bytes of the database file
+     */
+    getDatabaseSize: () => Promise<{bytesUsed: number; bytesRemaining: number}>;
+
+    /**
+     * @param  onStorageKeyChanged Storage synchronization mechanism keeping all opened tabs in sync
```
Comment (fabioh8010) [inline lib/storage/providers/types.ts:75]:
```suggestion
     * @param onStorageKeyChanged Storage synchronization mechanism keeping all opened tabs in sync
```

File: lib/storage/WebStorage.ts (hunk 4 above)
Comment (blazejkustra) [reply lib/storage/WebStorage.ts:17]: `localStorage.removeItem` only takes one argument

File: lib/OnyxCache.ts (hunk 3 above)
Comment (blazejkustra) [reply lib/OnyxCache.ts:25]: Looking at the usage I think it is unnecessary, but for sake of mind I can leave ii 👍

File: lib/OnyxCache.ts
Comment (blazejkustra) [reply lib/OnyxCache.ts:1]: I adjusted the code, now it's more accurate

File: lib/storage/providers/types.ts (hunk 7 above)
Comment (blazejkustra) [reply lib/storage/providers/types.ts:4]: lib.dom.d.ts 😄

File: lib/OnyxCache.ts
Comment (fabioh8010) [reply lib/OnyxCache.ts:1]: Thanks! There is still this pending item

> Could you add back the comments of all functions here?

File: react-native-onyx-2.0.4.tgz
Comment (fabioh8010) [inline react-native-onyx-2.0.4.tgz:1]: 👀

**Review by fabioh8010 — APPROVED** [review]

**Comment by blazejkustra** [comment]
@pecanoro All yours!

File: lib/utils.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+type EmptyObject = Record<string, never>;
+type EmptyValue = EmptyObject | null | undefined;
+
+function isEmptyObject<T>(obj: T | EmptyValue): obj is EmptyValue {
+    return typeof obj === 'object' && Object.keys(obj || {}).length === 0;
+}
+
+// Mostly copied from https://medium.com/@lubaka.a/how-to-remove-lodash-performance-improvement-b306669ad0e1
+
+function isMergeableObject(value: unknown): boolean {
```
Comment (pecanoro) [inline lib/utils.ts:13]: Can you explain for this function in plain English when the object is mergeable?

File: lib/storage/providers/SQLiteStorage.ts [hunk 10]
```diff
@@ -85,84 +58,45 @@ const provider = {
...
-    /**
-     * Returns all keys available in storage
-     * @returns {Promise<String[]>}
-     */
     getAllKeys: () =>
         db.executeAsync('SELECT record_key FROM keyvaluepairs;').then(({rows}) => {
             // eslint-disable-next-line no-underscore-dangle
-            const result = _.map(rows._array, (row) => row.record_key);
-            return result;
+            const result = rows?._array.map((row) => row.record_key);
```
Comment (pecanoro) [inline lib/storage/providers/SQLiteStorage.ts:75]: What happens here if rows in null? Is it returning the same as before?

**Review by pecanoro — CHANGES_REQUESTED** [review]

File: .eslintrc.js [hunk 11]
```diff
@@ -53,6 +53,20 @@ module.exports = {
                 '@typescript-eslint/array-type': ['error', {default: 'array-simple'}],
                 '@typescript-eslint/consistent-type-definitions': 'off',
                 'rulesdir/no-multiple-onyx-in-file': 'off',
+                'valid-jsdoc': 'off',
+                'rulesdir/prefer-import-module-contents': 'off',
+                'es/no-optional-chaining': 'off',
+                'es/no-nullish-coalescing-operators': 'off',
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply .eslintrc.js:59]: It was discussed before we migrated NewDot, so it's safe to assume the same for onyx

File: lib/utils.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    const nonNullObject = value != null ? typeof value === 'object' : false;
+    return nonNullObject && Object.prototype.toString.call(value) !== '[object RegExp]' && Object.prototype.toString.call(value) !== '[object Date]' && !Array.isArray(value);
+}
+
+function mergeObject<TTarget extends unknown[] | Record<string, unknown>>(target: TTarget, source: TTarget, shouldRemoveNullObjectValues = true): TTarget {
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply lib/utils.ts]: Done! 😄

File: lib/storage/providers/SQLiteStorage.ts [hunk 13]
```diff
@@ -17,65 +20,35 @@ db.execute('PRAGMA CACHE_SIZE=-20000;');
...
-    /**
-     * Get the value of a given key or return `null` if it's not available in storage
-     * @param {String} key
-     * @return {Promise<*>}
-     */
+const provider: StorageProvider = {
     getItem(key) {
         return db.executeAsync('SELECT record_key, valueJSON FROM keyvaluepairs WHERE record_key = ?;', [key]).then(({rows}) => {
-            if (rows.length === 0) {
+            if (rows?.length === 0) {
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply lib/storage/providers/SQLiteStorage.ts:26]: My guess is that it's never null. I tend to avoid logic changes when migration to TS to avoid any regressions

File: lib/storage/providers/SQLiteStorage.ts (hunk 10 above)
Comment (blazejkustra) [reply lib/storage/providers/SQLiteStorage.ts:75]: You are right, before it was an empty array, after this change it resulted in undefined, thank you! Adjusted in last commit.

File: lib/storage/providers/SQLiteStorage.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (pecanoro) [reply lib/storage/providers/SQLiteStorage.ts:26]: But aren't we changing the logic by adding the `?`? 😄

[the remark this answers is not in the document]
Comment (blazejkustra) [reply lib/storage/providers/SQLiteStorage.ts:26]: Actually logic is the same, just more safe:
Screenshot 2024-02-15 at 19 36 16

But actually why not add the condition you suggested! Changed in the last commit

**Review by pecanoro — APPROVED** [review]

**Comment by pecanoro** [comment]
Thank you for your patience!
